### PR TITLE
Allow downloading many artifacts and support unarchiving

### PR DESCRIPTION
**Pull request by dadgar**
This PR:
* changes the config of the `qemu`, `java`, `exec`, `raw_exec` drivers to remove artifacts sources
* adds a top level object `artifact` that can be specified many times.
* allows for unarchiving of artifacts

### CODE REVIEW

File: api/tasks.go [hunk 1]
```diff
@@ -92,6 +92,12 @@ type Task struct {
 	Meta        map[string]string
 	KillTimeout time.Duration
 	LogConfig   *LogConfig
+	Artifacts   []*TaskArtifact
+}
+
+type TaskArtifact struct {
```
Comment (diptanu) [inline api/tasks.go]: Comment?

File: client/driver/executor/executor.go [hunk 2]
```diff
@@ -280,3 +278,24 @@ func (e *UniversalExecutor) configureTaskDir() error {
 	e.cmd.Dir = taskDir
 	return nil
 }
+
+// makeExecutablePosix makes the given file executable for root,group,others.
+func (e *UniversalExecutor) makeExecutablePosix(binPath string) error {
+	fi, err := os.Stat(binPath)
```
Comment (diptanu) [inline client/driver/executor/executor.go]: Would this work if the binPath is not absolute?

File: client/driver/java.go [hunk 3]
```diff
@@ -160,7 +149,7 @@ func (d *JavaDriver) Start(ctx *ExecContext, task *structs.Task) (DriverHandle,
 		Cmd: exec.Command(bin, "executor", pluginLogFile),
 	}
 
-	exec, pluginClient, err := createExecutor(pluginConfig, d.config.LogOutput, d.config)
+	execImpl, pluginClient, err := createExecutor(pluginConfig, d.config.LogOutput, d.config)
```
Comment (diptanu) [inline client/driver/java.go]: We get an interface back not an impl.

File: client/driver/qemu.go [hunk 4]
```diff
@@ -98,36 +96,19 @@ func (d *QemuDriver) Start(ctx *ExecContext, task *structs.Task) (DriverHandle,
...
+	vmPath := driverConfig.ImagePath
+	if vmPath == "" {
+		return nil, fmt.Errorf("image_path must be set")
 	}
+	vmID := filepath.Base(vmPath)
 
 	// Get the tasks local directory.
-	taskDir, ok := ctx.AllocDir.TaskDirs[d.DriverContext.taskName]
+	taskName := d.DriverContext.taskName
+	taskDir, ok := ctx.AllocDir.TaskDirs[taskName]
```
Comment (diptanu) [inline client/driver/qemu.go]: Inline taskName?

File: jobspec/parse_test.go [hunk 5]
```diff
@@ -128,6 +128,20 @@ func TestParse(t *testing.T) {
 									MaxFiles:      10,
 									MaxFileSizeMB: 100,
 								},
+								Artifacts: []*structs.TaskArtifact{
```
Comment (diptanu) [inline jobspec/parse_test.go]: Can we add a test where we pass invalid keys to Artifact options and have the parser fail?

File: nomad/structs/structs.go [hunk 6]
```diff
@@ -1662,6 +1677,72 @@ func (t *Task) FindHostAndPortFor(portLabel string) (string, int) {
 	return "", 0
 }
 
+// Validate is used to sanity check a task group
```
Comment (diptanu) [inline nomad/structs/structs.go]: Incorrect comment

File: client/driver/executor/executor.go (hunk 2 above)
Comment (dadgar) [inline client/driver/executor/executor.go]: Yeah it would be relative to Nomads cwd tho

File: jobspec/parse_test.go (hunk 5 above)
Comment (dadgar) [inline jobspec/parse_test.go]: Done

**Pull request merged**